cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MessagingMetricsProcessorService.java

```diff
@@ -85,9 +88,12 @@
   private final MetricStore metricStore;
   private final Map<String, String> metricsContextMap;
   private final int fetcherLimit;
-  private final ConcurrentLinkedDeque<MetricValues> records;
+  private final int queueSize;
+  private final BlockingQueue<MetricValues> records;
   private final ConcurrentMap<TopicIdMetaKey, byte[]> topicMessageIds;
+  private final ConcurrentMap<TopicIdMetaKey, Long> topicLatestTime;
```

Comment:
You should merge this two concurrent map into one, by using a private static class that contains both the message id and time.